// write-excel-file@2.3.10/commonjs/write/writeXlsxFileNode.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = writeXlsxFile;
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
var _os = _interopRequireDefault(require("os"));
var _stream = _interopRequireWildcard(require("stream"));
var _archive = _interopRequireDefault(require("./archive.js"));
var _getImageFileName = _interopRequireDefault(require("./getImageFileName.js"));
var _workbookXml = _interopRequireDefault(require("./files/workbook.xml.js"));
var _workbookXmlRels = _interopRequireDefault(require("./files/workbook.xml.rels.js"));
var _rels = _interopRequireDefault(require("./files/rels.js"));
var _Content_TypesXml = _interopRequireDefault(require("./files/[Content_Types].xml.js"));
var _drawingXml = _interopRequireDefault(require("./files/drawing.xml.js"));
var _drawingXmlRels = _interopRequireDefault(require("./files/drawing.xml.rels.js"));
var _sheetXmlRels = _interopRequireDefault(require("./files/sheet.xml.rels.js"));
var _sharedStringsXml = _interopRequireDefault(require("./files/sharedStrings.xml.js"));
var _stylesXml = _interopRequireDefault(require("./files/styles.xml.js"));
var _writeXlsxFileCommon = require("./writeXlsxFile.common.js");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, "default": e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _createForOfIteratorHelperLoose(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (t) return (t = t.call(r)).next.bind(t); if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var o = 0; return function () { return o >= r.length ? { done: !0 } : { done: !1, value: r[o++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
// This function doesn't use `async`/`await` in order to avoid adding `@babel/runtime` to `dependencies`.
// https://gitlab.com/catamphetamine/write-excel-file/-/issues/105
function writeXlsxFile(data) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
    filePath = _ref.filePath,
    buffer = _ref.buffer,
    sheetName = _ref.sheet,
    sheetNames = _ref.sheets,
    schema = _ref.schema,
    columns = _ref.columns,
    images = _ref.images,
    headerStyle = _ref.headerStyle,
    getHeaderStyle = _ref.getHeaderStyle,
    fontFamily = _ref.fontFamily,
    fontSize = _ref.fontSize,
    orientation = _ref.orientation,
    stickyRowsCount = _ref.stickyRowsCount,
    stickyColumnsCount = _ref.stickyColumnsCount,
    showGridLines = _ref.showGridLines,
    rightToLeft = _ref.rightToLeft,
    dateFormat = _ref.dateFormat;
  var _generateSheets = (0, _writeXlsxFileCommon.generateSheets)({
      data: data,
      sheetName: sheetName,
      sheetNames: sheetNames,
      schema: schema,
      columns: columns,
      images: images,
      headerStyle: headerStyle,
      getHeaderStyle: getHeaderStyle,
      fontFamily: fontFamily,
      fontSize: fontSize,
      orientation: orientation,
      stickyRowsCount: stickyRowsCount,
      stickyColumnsCount: stickyColumnsCount,
      showGridLines: showGridLines,
      rightToLeft: rightToLeft,
      dateFormat: dateFormat
    }),
    sheets = _generateSheets.sheets,
    getSharedStrings = _generateSheets.getSharedStrings,
    getStyles = _generateSheets.getStyles;
  return createDirectories().then(function (_ref2) {
    var rootPath = _ref2.rootPath,
      xlPath = _ref2.xlPath,
      mediaPath = _ref2.mediaPath,
      drawingsPath = _ref2.drawingsPath,
      drawingsRelsPath = _ref2.drawingsRelsPath,
      relsPath = _ref2.relsPath,
      worksheetsPath = _ref2.worksheetsPath,
      worksheetsRelsPath = _ref2.worksheetsRelsPath;
    var promises = [writeFile(_path["default"].join(relsPath, 'workbook.xml.rels'), (0, _workbookXmlRels["default"])({
      sheets: sheets
    })), writeFile(_path["default"].join(xlPath, 'workbook.xml'), (0, _workbookXml["default"])({
      sheets: sheets,
      stickyRowsCount: stickyRowsCount,
      stickyColumnsCount: stickyColumnsCount
    })), writeFile(_path["default"].join(xlPath, 'styles.xml'), (0, _stylesXml["default"])(getStyles())), writeFile(_path["default"].join(xlPath, 'sharedStrings.xml'), (0, _sharedStringsXml["default"])(getSharedStrings()))];
    for (var _iterator = _createForOfIteratorHelperLoose(sheets), _step; !(_step = _iterator()).done;) {
      var _step$value = _step.value,
        id = _step$value.id,
        _data = _step$value.data,
        _images = _step$value.images;
      promises.push(writeFile(_path["default"].join(worksheetsPath, "sheet".concat(id, ".xml")), _data));
      promises.push(writeFile(_path["default"].join(worksheetsRelsPath, "sheet".concat(id, ".xml.rels")), (0, _sheetXmlRels["default"])({
        id: id,
        images: _images
      })));
      if (_images) {
        promises.push(writeFile(_path["default"].join(drawingsPath, "drawing".concat(id, ".xml")), (0, _drawingXml["default"])({
          images: _images
        })));
        promises.push(writeFile(_path["default"].join(drawingsRelsPath, "drawing".concat(id, ".xml.rels")), (0, _drawingXmlRels["default"])({
          images: _images,
          sheetId: id
        })));
        // Copy images to `xl/media` folder.
        for (var _iterator2 = _createForOfIteratorHelperLoose(_images), _step2; !(_step2 = _iterator2()).done;) {
          var image = _step2.value;
          var imageContentReadableStream = getReadableStream(image.content);
          var imageFilePath = _path["default"].join(mediaPath, (0, _getImageFileName["default"])(image, {
            sheetId: id,
            sheetImages: _images
          }));
          promises.push(writeFileFromStream(imageFilePath, imageContentReadableStream));
        }
      }
    }
    return Promise.all(promises).then(function () {
      var archive = createArchive(filePath, {
        sheets: sheets,
        images: images,
        xlDirectoryPath: xlPath
      });
      if (filePath) {
        // Doesn't return anything.
        return archive.getPromise().then(function () {
          return removeDirectoryWithLegacyNodeVersionsSupport(rootPath);
        }).then(function () {
          // Doesn't return anything.
        });
      } else if (buffer) {
        // Returns a `Buffer`.
        return streamToBuffer(archive.getStream());
      } else {
        // Returns a readable `Stream`.
        return archive.getStream();
      }
    });
  });
}

// Creates a `*.zip` archive from Excel data and returns a readable `Stream`.
function createArchive(outputFilePath, _ref3) {
  var sheets = _ref3.sheets,
    images = _ref3.images,
    xlDirectoryPath = _ref3.xlDirectoryPath;
  // I dunno why `Archive` class is used here instead of something like `JSZip`.
  // `JSZip` is already used in `writeXlsxFileBrowser.js`, so maybe it would've also fit here.
  // In that case, `Archive` class could potentially be removed.
  var archive = new _archive["default"](outputFilePath);
  archive.directory(xlDirectoryPath, 'xl');
  archive.append(_rels["default"], '_rels/.rels');
  archive.append((0, _Content_TypesXml["default"])({
    sheets: sheets,
    images: images
  }), '[Content_Types].xml');
  archive.write();
  return archive;
}

// According to Node.js docs:
// https://nodejs.org/api/fs.html#fswritefilefile-data-options-callback
// `contents` argument could be of type:
// * string — File path
// * Buffer
// * TypedArray
// * DataView
function writeFile(path, contents) {
  return new Promise(function (resolve, reject) {
    _fs["default"].writeFile(path, contents, 'utf-8', function (error) {
      if (error) {
        return reject(error);
      }
      resolve();
    });
  });
}
function createDirectory(path) {
  return new Promise(function (resolve, reject) {
    _fs["default"].mkdir(path, function (error) {
      if (error) {
        return reject(error);
      }
      resolve(path);
    });
  });
}
function createTempDirectory() {
  return new Promise(function (resolve, reject) {
    _fs["default"].mkdtemp(_path["default"].join(_os["default"].tmpdir(), 'write-excel-file-'), function (error, directoryPath) {
      if (error) {
        return reject(error);
      }
      resolve(directoryPath);
    });
  });
}
function removeDirectoryWithLegacyNodeVersionsSupport(path) {
  if (_fs["default"].rm) {
    return removeDirectory(path);
  } else {
    removeDirectoryLegacySync(path);
    return Promise.resolve();
  }
}

// `fs.rm()` is available in Node.js since `14.14.0`.
function removeDirectory(path) {
  return new Promise(function (resolve, reject) {
    _fs["default"].rm(path, {
      recursive: true,
      force: true
    }, function (error) {
      if (error) {
        return reject(error);
      }
      resolve();
    });
  });
}

// For Node.js versions below `14.14.0`.
function removeDirectoryLegacySync(directoryPath) {
  var childNames = _fs["default"].readdirSync(directoryPath);
  for (var _iterator3 = _createForOfIteratorHelperLoose(childNames), _step3; !(_step3 = _iterator3()).done;) {
    var childName = _step3.value;
    var childPath = _path["default"].join(directoryPath, childName);
    var stats = _fs["default"].statSync(childPath);
    if (childPath === '.' || childPath === '..') {
      // Skip.
    } else if (stats.isDirectory()) {
      // Remove subdirectory recursively.
      removeDirectoryLegacySync(childPath);
    } else {
      // Remove file.
      _fs["default"].unlinkSync(childPath);
    }
  }
  _fs["default"].rmdirSync(directoryPath);
}

// https://stackoverflow.com/a/67729663
function streamToBuffer(stream) {
  return new Promise(function (resolve, reject) {
    var chunks = [];
    stream.on('data', function (chunk) {
      return chunks.push(chunk);
    });
    stream.on('end', function () {
      return resolve(Buffer.concat(chunks));
    });
    stream.on('error', reject);
  });
}
function copyFile(fromPath, toPath) {
  return new Promise(function (resolve, reject) {
    _fs["default"].copyFile(fromPath, toPath, function (error) {
      if (error) {
        return reject(error);
      }
      resolve();
    });
  });
}
function getReadableStream(source) {
  if (source instanceof _stream["default"]) {
    return source;
  }
  if (source instanceof Buffer) {
    return _stream.Readable.from(source);
  }
  if (typeof source === 'string') {
    return _fs["default"].createReadStream(source);
  }
  throw new Error('Unsupported content source: couldn\'t convert it to a readable stream');
}
function writeFileFromStream(filePath, readableStream) {
  var writableStream = _fs["default"].createWriteStream(filePath);
  readableStream.pipe(writableStream);
  return new Promise(function (resolve) {
    return writableStream.on('finish', resolve);
  });
}
function createDirectories() {
  // There doesn't seem to be a way to just append a file into a subdirectory
  // in `archiver` library, hence using a hacky temporary directory workaround.
  // https://www.npmjs.com/package/archiver
  return createTempDirectory().then(function (rootPath) {
    var xlPath = _path["default"].join(rootPath, 'xl');
    var mediaPath = _path["default"].join(xlPath, 'media');
    var drawingsPath = _path["default"].join(xlPath, 'drawings');
    var drawingsRelsPath = _path["default"].join(drawingsPath, '_rels');
    var relsPath = _path["default"].join(xlPath, '_rels');
    var worksheetsPath = _path["default"].join(xlPath, 'worksheets');
    var worksheetsRelsPath = _path["default"].join(worksheetsPath, '_rels');
    var directories = [xlPath, mediaPath, drawingsPath, drawingsRelsPath, relsPath, worksheetsPath, worksheetsRelsPath];
    var createAllDirectoriesPromise = directories.reduce(function (promise, directory) {
      return promise.then(function () {
        return createDirectory(directory);
      });
    }, Promise.resolve());
    return createAllDirectoriesPromise.then(function () {
      return {
        rootPath: rootPath,
        xlPath: xlPath,
        mediaPath: mediaPath,
        drawingsPath: drawingsPath,
        drawingsRelsPath: drawingsRelsPath,
        relsPath: relsPath,
        worksheetsPath: worksheetsPath,
        worksheetsRelsPath: worksheetsRelsPath
      };
    });
  });
}
